src/lib/mina_ledger/sparse_ledger.ml

```diff
@@ -33,36 +33,37 @@ let of_ledger_subset_exn (oledger : Ledger.t) keys =
                loc := next_location_exn !loc ;
                !loc )
     in
-    let merkle_paths =
-      Ledger.merkle_path_batch oledger (empty_locations @ non_empty_locations)
-    in
-    List.split_n merkle_paths num_new_accounts
+    let paths = path_query oledger (empty_locations @ non_empty_locations) in
+    List.split_n paths num_new_accounts
+  in
+  let process_location sl key = function
+    | Some _, (_, Some account) :: accs, path :: ne_paths, epaths ->
+        (path_add sl path key account, accs, ne_paths, epaths)
+    | None, accs, ne_paths, path :: epaths ->
+        (path_add sl path key Account.empty, accs, ne_paths, epaths)
+    | Some _, (_, None) :: _, _, _ ->
+        failwith
+          "of_ledger_subset_exn: account not found for location returned by \
+           location_of_account_batch"
+    | _ ->
+        failwith "of_ledger_subset_exn: mismatched lengths"
   in
   let sl, _, _, _ =
     List.fold locations
-      ~init:(of_ledger_root oledger, accounts, merkle_paths, empty_merkle_paths)
-      ~f:(fun (sl, accounts, merkle_paths, empty_merkle_paths) (key, location) ->
-        match location with
-        | Some _loc -> (
-            match (accounts, merkle_paths) with
-            | (_, account) :: rest, merkle_path :: rest_merkle_paths ->
-                let sl =
-                  add_path sl merkle_path key (Option.value_exn account)
-                in
-                (sl, rest, rest_merkle_paths, empty_merkle_paths)
-            | _ ->
-                failwith "unexpected number of non empty accounts" )
-        | None ->
-            let path = List.hd_exn empty_merkle_paths in
-            let sl = add_path sl path key Account.empty in
-            (sl, accounts, merkle_paths, List.tl_exn empty_merkle_paths) )
+      ~init:(of_ledger_root oledger, accounts, non_empty_paths, empty_paths)
+      ~f:(fun (sl, accs, ne_paths, epaths) (key, mloc) ->
+        process_location sl key (mloc, accs, ne_paths, epaths) )
   in
   Debug_assert.debug_assert (fun () ->
       [%test_eq: Ledger_hash.t]
         (Ledger.merkle_root oledger)
         ((merkle_root sl :> Random_oracle.Digest.t) |> Ledger_hash.of_hash) ) ;
   sl
 
+let of_ledger_subset_exn =
+  of_ledger_subset_exn_impl ~path_query:Ledger.wide_merkle_path_batch
```

Comment:
Why do we have a separate impl helper for this function? It seems we only call it in this one location.